openmc/tallies.py

```diff
@@ -3522,6 +3527,18 @@ def _create_mesh_subelements(self, root_element):
                         root_element.append(f.mesh.to_xml_element())
                         already_written.add(f.mesh)
 
+    def _create_filter_subelements(self, root_element):
+        already_written = dict()
+        for tally in self:
+            for f in tally.filters:
+                if f not in already_written:
+                    root_element.append(f.to_xml_element())
+                    already_written[f] = f.id
+                else:
+                    # Set the IDs of identical filters with different
+                    # user-defined IDs to the same value
+                    f.id = already_written[f]
```

Comment:
Filters are replicated if multiple `MGXS` objects are created with the same domain or energy groups. It looks like each `MGXS` object will contain only unique filters, but identical filters with different IDs can be created across objects because they’re being initialized again with a different auto-generated ID. This happens in `Library.build_library()`: a new `MGXS` object is initialized for each combination of domain and reaction types, each with the same energy groups.

Consider, as an example, test_mgxs_library_ce_to_mg.py: there are five reaction types and a material domain type with three materials. This means when `Library.build_library()` is called, the same energy filter is replicated 15 times, an energyout filter is replicated 9 times, and each of three material filters is replicated five times, for a total of 39 filters out of which 5 are unique.